TNL 3294 - Prevent double click creating multiple teams

efischer19 opened the pull request:
## [TNL-3294](https://openedx.atlassian.net/browse/TNL-3294)

Preventing issue of team being created twice on double-click

### Testing
- [x] manual testing (try to trigger bug many times, watch all attempts fail. Both locally and on sandbox)

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @andy-armstrong 
- [x] Code review: @cahrens 

FYI: @explorerleslie 
 
### Post-review
- [x] Squash commits

efischer19 commented:
jenkins run js

efischer19 commented:
Can you give this a look @andy-armstrong @cahrens?

I think a better solution would be to disable the button after one click and re-enable it when appropriate, but I couldn't figure out how to get that solution working without adding a jQuery dependency. Will this `_.debounce()` approach work for us?

Reference: http://eng.wealthfront.com/2012/12/using-underscorejss-debounce-to-filter.html

efischer19 commented:
jenkins run python

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
jenkins run js

efischer19 commented:
jenkins run lettuce

cahrens commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -145,7 +144,10 @@
-                        });
+                        })
+                    ViewUtils.disableElementWhileRunning($(".action-primary"), function() {
+                        return deferred.promise();
```
Don't you have the deferred.promise and deferred.resolve reversed? I would expect the method to return deferred.promise(), and disableElementWhileRunning to do the resolve inside of its function.

Is disableElementWhileRunning actually used anywhere in the codebase currently?

efischer19 replied:
The only place I found was here: https://github.com/edx/edx-platform/blob/d38da773ea2ff4b139e2c6b6d61e02681584fcab/common/static/common/js/spec/components/view_utils_spec.js#L15

I ended up removing the `resolve()` call entirely, that was throwing errors in js tests. That still seems to work.

So yeah, you're right - the anonymous method here returns a promise, `disableElementWhileRunning` does the resolve internally.

andy-armstrong replied:
I find this hard to read when the function in question is just a promise. I think it would be clearer if you moved the ``save`` call inside and then you don't need a variable too:

    ViewUtils.disableElementWhileRunning($(".action-primary"), function() {
        return this.teamModel.save(data, saveOptions)
            ...
    });

efischer19 commented:
jenkins run python

efischer19 commented:
jenkins run lettuce

efischer19 commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -6,15 +6,16 @@
 
                 events: {
-                    'click .action-primary': 'createOrUpdateTeam',
+                    'click .action-primary': function(event) {this.disableRun(event, this.createOrUpdateTeam);},
```
@andy-armstrong, what about an approach like this, where the event handler is a generalized function (in `ViewUtils` I'd imagine) that wraps the actual function?

andy-armstrong replied:
Does that work? I was trying to see yesterday if you could pass a function instead of a string. You could make it shorter like this:

    ViewUtils.onButtonClick(this.createOrUpdateTeam)

to mirror how ``_.bind`` works. That function would in turn return a function that does the logic you have. I can't think of a great name for such a function, because of all the things it does for you.

efischer19 replied:
It "works", to the point that it tries to call this, but fails with `operation is not a function` inside of disableElementWhileRunning.

I had to wrap everything here in an anonymous function due to `this.createOrUpdateTeam` not being defined at the time this event map is declared.

efischer19 replied:
![screen shot 2015-09-18 at 3 19 25 pm](https://cloud.githubusercontent.com/assets/7373924/9968813/b604b28a-5e18-11e5-8e75-225e33c6f651.png)

This is where I'm running into issues. It is completing successfully though, since in the non-failure case we're redirecting anyways.

andy-armstrong replied:
Ah, makes sense. You could pass it in as a string, and then access it from ``this`` when it is called. That must be why the events are strings in the first place.

andy-armstrong replied:
The value of operation is a promise, not the function. Your disableRun function is evaluating the function and getting the promise and passing that to ``disableElementWhileRunning``. As it is defined, it accepts a function which when evaluated will return a promise. 

Coding with promises is always like this...

andy-armstrong replied:
This is much better now. I think you should move the anonymous function into ``disableRun`` and then you can make this line just:

    'click .action-primary': this.disableRun('createOrUpdateTeam')

I believe you have to make it a string as you mentioned above, because the field isn't available at the time of creation. As you suggested, it would be even better to have this be ``ViewUtils.disableRun``. In fact, I'd go with ``ViewUtils.withDisabledButton`` or ``ViewUtils.withDisabledElement`` to make it clearer what it is doing for you.

andy-armstrong replied:
@efischer19 Sorry to do this at this point, but the solution you have still doesn't end up the way I was suggesting in this thread here. I think it is important that this can be added just by wrapping the event, and that it doesn't require an extra function in order to make things work. The trick as I mentioned is to move the anonymous function into the helper method.

I've pushed up a commit that converts what you have to the syntax that I was proposing (my commit is currently 537280f373218d6e466cc3798f489dc0ff483ca2). It is now:

    'click .action-primary': ViewUtils.withDisabledElement('createOrUpdateTeam')

Let's discuss first thing in the morning and make sure that I'm not missing something.

efischer19 commented:
Alright, this is looking much better now. How does everyone feel about using this general approach more frequently in the codebase? We'd probably want to move `disableRun` into ViewUtils, then we could have the anonymous function in the `events` map any time it's needed.

Or is that too complex? I'm leaning towards doing it, but open to discussion.

@cahrens - this is ready for you to look at once more if you want to weigh in.

efischer19 commented:
*sigh*

Nope, js test failures

efischer19 commented:
jenkins run js

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -124,10 +133,9 @@
                     if (validationResult.status === false) {
                         this.showMessage(validationResult.message, validationResult.srMessage);
-                        return;
+                        return $().promise();
```
I've always used:

    $.Deferred().resolve().promise()

I don't know which is better idiomatically, but yours is shorter! 

http://stackoverflow.com/questions/25032549/use-of-dummy-and-empty-jquery-promises

efischer19 replied:
Haha, that's the exact same link I found this morning when looking at how to fix things. I suppose we'll go with the shorter version then!

efischer19 commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -6,15 +6,16 @@
 
                 events: {
-                    'click .action-primary': 'createOrUpdateTeam',
+                    'click .action-primary': function(event) {ViewUtils.withDisabledElement(event, 'createOrUpdateTeam', this);},
```
@andy-armstrong is there some way I can get this to be just `ViewUtils.withDisabledElement`, without the anonymous function wrapper?

cahrens replied:
I would recommend making a method in this class that calls ViewUtils.withDisabledElement. That way you can just specify the name of the method in the manner of the other event handlers.

cahrens commented on common/static/common/js/components/utils/view_utils.js:
```diff
@@ -7,8 +7,8 @@
-            runOperationShowingMessage, disableElementWhileRunning, getScrollOffset, setScrollOffset,
-            setScrollTop, redirect, reload, hasChangedAttributes, deleteNotificationHandler,
+            runOperationShowingMessage, withDisabledElement, disableElementWhileRunning,
+            getScrollOffset, setScrollOffset, setScrollTop, redirect, reload, hasChangedAttributes, deleteNotificationHandler,
```
If disableElementWhileRunning isn't used anywhere (except its unit test), and it wasn't useful for this case, I would propose that we delete it. But I don't know the issues about why it couldn't be used... so I leave that up to your judgement, @efischer19, as to whether or not the method is potentially useful.

cahrens commented on common/static/common/js/components/utils/view_utils.js:
```diff
@@ -88,6 +88,23 @@
+         * paradigm is designed to be used in Backbone event maps where
+         * multiple events firing simultaneously is not desired.
+         * @param event the event whose handler needs to only run once
+         * @param functionName the function to execute, as a string.
```
Arguments need to be updated.

efischer19 replied:
Check

cahrens commented on common/static/common/js/components/utils/view_utils.js:
```diff
@@ -88,6 +88,23 @@
+        withDisabledElement = function(event, callback, self) {
+            event.preventDefault();
+            disableElementWhileRunning($(event.currentTarget), function() {
+                return self[callback].apply(self, []);
```
Can you add a comment about what the second argument is ([])?

cahrens commented on common/static/common/js/components/utils/view_utils.js:
```diff
@@ -88,6 +88,23 @@
+         * @param context the context in which to run the function
+         * @param arguments other arguments to pass into the function
+         */
+        withDisabledElement = function(event, callback, self) {
```
I'd like to see a unit test for withDisabledElement.

efischer19 commented:
Sandbox is out-of-date, and I still need to write a unit test

efischer19 commented:
@cahrens, I believe I've addressed all your comments. Does the unit test I wrote cover things well enough, or do you think we ought to also add some type of test coverage for unintended double-click behavior?

cahrens commented on lms/djangoapps/teams/static/teams/js/views/edit_team.js:
```diff
@@ -6,15 +6,20 @@
                 maxTeamNameLength: 255,
                 maxTeamDescriptionLength: 300,
 
+                submitNoDoubleCLick: function(event) {
```
Typo: capitalized "L".

cahrens replied:
I would move this function definition down below in the file, where createOrUpdateTeam is. It's odd to have any functions above the event handling declaration and initialize methods.

efischer19 replied:
Nice catch - I had meant to ask about the placement of that. Since `createOrUpdateTeam` is hard-coded in, I think down near there is fine. Should be fixed now.

cahrens commented on common/static/common/js/spec/components/view_utils_spec.js:
```diff
@@ -17,6 +17,24 @@ define(["jquery", "underscore", "common/js/components/utils/view_utils", "common
+
+                it("uses withDisabledElement wrapper to disable element while running a Backbone event handler", function() {
+                    var link,
+                        event = {var currentTarget = link;},
```
This is invalid JS, and I suspect the test is not actually being run.

cahrens commented:
:+1:

efischer19 commented:
@andy-armstrong This has been a noisy PR, but I think it's now in the home stretch. Would you mind looking over it once more before I merge?

efischer19 commented:
Thanks for the fix @andy-armstrong, that's what I had been trying to get working for a general-purpose solution.

The revert-and-commit I just made should have all of your changes, but without the additional whitespace changes. Everything look okay to you? It passes jasmine tests and manual tests run locally.

FYI @cahrens the changes made since your previous thumbs are all here: https://github.com/edx/edx-platform/commit/2d8b98c17298fe1b2c56a5f5bc6fa2439f8bd33a

cahrens commented:
:+1: on the changes-- make sure to squash commits before merging.

andy-armstrong commented:
:+1: Thanks for sticking with this one! Feel free to squash my commit away so that this is just a single commit.
